File: Chapter.03/FunctionalDataStructures.fs/WeakMemoization.fs
module WeakMemoization

open System
open System.Collections.Concurrent
open System.Runtime.CompilerServices


let memoizeWeakWithTtl (func :'a->'b) ttl =
    let keyStore = ConcurrentDictionary<int, 'a>()
    let reduceKey (obj: 'a) =
        let oldObj = keyStore.GetOrAdd(obj.GetHashCode(), obj)
        if obj.Equals(oldObj) then oldObj else obj

    let cache = ConditionalWeakTable<'a, 'b * DateTime>()
    let factoryFunc =
        ConditionalWeakTable<'a, 'b * DateTime>
            .CreateValueCallback(
                fun key -> func key, DateTime.Now + ttl)

    fun (arg:'a) ->
        let key = reduceKey arg

        let (value, due) = cache.GetValue(key, factoryFunc)
        if (due < DateTime.Now)
        then
            let newValue = factoryFunc.Invoke(key)
            cache.Remove(key) |> ignore
            cache.Add(key, newValue)
            fst newValue
        else value


let example() =
    let greating name =
        sprintf "Warm greetings %s, the time is %s"
            name (DateTime.Now.ToString("hh:mm:ss"))

    let greetingMemoize =
        memoizeWeakWithTtl greating (TimeSpan.FromSeconds(2.0))

    printfn "%s" <| greetingMemoize("Richard")
    System.Threading.Thread.Sleep(1500)
    printfn "%s" <| greetingMemoize("_Richard".Substring(1))
    System.Threading.Thread.Sleep(1500)
    printfn "%s" <| greetingMemoize("_Richard".Substring(1))

File: Chapter.02/FunctionalTechniques.CSharp/WeakMemoization.cs
using System;
using System.Collections.Concurrent;
using System.Runtime.CompilerServices;
using System.Threading;

namespace FunctionalTechniques.CSharp
{
    public static partial class Memoization
    {
        public static Func<T, R> MemoizeWeakWithTtl<T, R>(Func<T, R> func, TimeSpan ttl)
            where T : class, IEquatable<T>
            where R : class
        {
            var keyStore = new ConcurrentDictionary<int, T>();

            T ReduceKey(T obj)
            {
                var oldObj = keyStore.GetOrAdd(obj.GetHashCode(), obj);
                return obj.Equals(oldObj) ? oldObj : obj;
            }

            var cache = new ConditionalWeakTable<T, Tuple<R, DateTime>>();

            Tuple<R, DateTime> FactoryFunc(T key)
            {
                return new Tuple<R, DateTime>(func(key), DateTime.Now + ttl);
            }

            return arg =>
            {
                var key = ReduceKey(arg);
                var value = cache.GetValue(key, FactoryFunc);
                if (value.Item2 >= DateTime.Now)
                    return value.Item1;
                value = FactoryFunc(key);
                cache.Remove(key);
                cache.Add(key, value);
                return value.Item1;
            };
        }


        public static void Example()
        {
            string Greeting(string name)
            {
                return $"Warm greetings {name}, the time is {DateTime.Now:hh:mm:ss}";
            }


            var greetingMemoize = MemoizeWeakWithTtl<string, string>(Greeting, TimeSpan.FromDays(1.0));

            Console.WriteLine(greetingMemoize("Richard"));
            Thread.Sleep(1500);
            Console.WriteLine(greetingMemoize("_Richard".Substring(1)));
        }
    }
}